pkg/controllers/machine/lifecycle/launch.go

```diff
@@ -0,0 +1,147 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package lifecycle
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/patrickmn/go-cache"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/samber/lo"
+	v1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	"github.com/aws/karpenter-core/pkg/metrics"
+	"github.com/aws/karpenter-core/pkg/scheduling"
+)
+
+type Launch struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+	cache         *cache.Cache // exists due to eventual consistency on the cache
+}
+
+func (l *Launch) Reconcile(ctx context.Context, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	if machine.StatusConditions().GetCondition(v1alpha5.MachineLaunched).IsTrue() {
+		return reconcile.Result{}, nil
+	}
+
+	var err error
+	var created *v1alpha5.Machine
+
+	if ret, ok := l.cache.Get(client.ObjectKeyFromObject(machine).String()); ok {
+		created = ret.(*v1alpha5.Machine)
+	} else if _, ok := machine.Annotations[v1alpha5.MachineLinkedAnnotationKey]; ok {
+		created, err = l.linkMachine(ctx, machine)
+	} else {
+		created, err = l.launchMachine(ctx, machine)
+	}
+	// Either the machine launch/linking failed or the machine was deleted due to InsufficientCapacity/NotFound
+	if err != nil || created == nil {
+		return reconcile.Result{}, err
```

Comment:
Won't we reach here if we get the machine from the cache, but it's not linked or launched yet? Maybe I'm missing the exact flow, but I think in that case, we should requeue, but it seems like we wouldn't here.

Reply:
You're correct, but the only way something gets added to the cache is if it's linked or launched. The goal of the cache is to make sure that we don't relaunch if we either:
1. Fail to patch the status
2. Happen to re-queue too quickly and don't get the updated Machine version from the client-go cache